dependencies/windows/install-boost/install-boost.R

```diff
@@ -1,63 +1,119 @@
+   "scope_exit", "signals", "smart_ptr", "spirit", "string_algo",
+   "system", "test", "thread", "tokenizer", "type_traits", "typeof",
+   "unordered", "utility", "variant"
+)
 
 # construct paths of interest
 boost_filename <- basename(boost_url)
 boost_dirname <- tools::file_path_sans_ext(boost_filename)
-boost_zippath <- boost_filename
+boost_archive <- boost_filename
 
 if (!file.exists(boost_dirname)) {
    
-   section("Downloading Boost...")
-   
-   download(boost_url, destfile = boost_zippath)
-   if (!file.exists(boost_zippath))
-      fatal("Failed to download '%s'", boost_zippath)
+   # download boost if we don't have it
+   if (!file.exists(boost_archive)) {
+      section("Downloading Boost...")
+      download(boost_url, destfile = boost_archive)
+      if (!file.exists(boost_archive))
+         fatal("Failed to download '%s'", boost_archive)
+   }
    
    # extract boost (be patient, this will take a while)
    progress("Extracting archive -- please wait a moment...")
-   exec("unzip", boost_filename)
+   progress("Feel free to get up and grab a coffee.")
+   exec("7z.exe", "x -mmt4", boost_filename)
 }
 
+# double-check that we generated the boost folder
 if (!file.exists(boost_dirname))
    fatal("'%s' doesn't exist (download / extract failed?)", boost_dirname)
 
+# clean up the path
 PATH$remove(toolchain_32bit)
 PATH$remove(toolchain_64bit)
 
+# enter boost folder
 enter(boost_dirname)
 
-# Bootstrap the boost build directory
+# remove any documentation folders (these cause
+# bcp to barf while trying to copy files)
+unlink("doc", recursive = TRUE)
+docs <- file.path("libs", list.files("libs"), "doc")
+invisible(lapply(docs, function(doc) {
+   if (file.exists(doc))
+      unlink(doc, recursive = TRUE)
+}))
+
+# bootstrap the boost build directory
 PATH$prepend(toolchain_32bit)
 section("Bootstrapping boost...")
-exec("cmd.exe /C call bootstrap.bat gcc")
+exec("cmd.exe", "/C call bootstrap.bat gcc")
+
+# create bcp executable (so we can create Boost
+# using a private namespace)
+exec("b2", "toolset=gcc -j 4 tools\\bcp")
+invisible(file.copy("dist/bin/bcp.exe", "bcp.exe"))
+
+# use bcp to copy boost into 'rstudio' sub-directory
+unlink("rstudio", recursive = TRUE)
+dir.create("rstudio")
+fmt <- "--namespace=rstudio_boost --namespace-alias %s config build rstudio"
+args <- sprintf(fmt, paste(boost_modules, collapse = " "))
+exec("bcp", args)
+
+# enter the 'rstudio' directory and re-bootstrap
+enter("rstudio")
+exec("cmd.exe", "/C call bootstrap.bat gcc")
 PATH$remove(toolchain_32bit)
 
-# Construct common arguments for 32bit, 64bit boost builds
-bjam_command <- function(bitness) {
+# construct common arguments for 32bit, 64bit boost builds
+b2_build_args <- function(bitness) {
+   
+   prefix <- file.path(output_dir, sprintf("boost%s", bitness))
+   cxxflags <- "-march=core2 -mtune=generic"
+   
+   unlink(prefix, recursive = TRUE)
+   
    paste(
-      "b2",
+      "-q",
```

Comment:
`-q` here just means 'stop on first error', so that the build dies more quickly on error.